go/types/encode_human_readable.go

```diff
@@ -9,12 +9,77 @@ import (
 	"fmt"
 	"io"
 	"strconv"
+	"sync"
 
 	"github.com/attic-labs/noms/go/d"
 	"github.com/attic-labs/noms/go/util/writers"
 	humanize "github.com/dustin/go-humanize"
 )
 
+// Clients can register a function to return a comment that will get appended
+// to the first line of encoded structs. For example, the noms DateTime struct
+// normally gets encoded as follows:
+//    lastRefresh: DateTime {
+//      secSinceEpoch: 1.501801626877e+09,
+//    }
+//
+// By registering a commenter function that returns a nicely formatted date,
+// the struct will be coded with a comment:
+//    lastRefresh: DateTime { // 2017-08-03T16:07:06-07:00
+//      secSinceEpoch: 1.501801626877e+09,
+//    }
+
+// Function type for commenter functions
+type HRSCommentFunc func(Struct) string
+
+var (
+	commenterRegistry = map[string]map[string]HRSCommentFunc{}
+	registryLock      sync.RWMutex
+)
+
+// RegisterHRSCommenter is called to with three arguments:
+//  typename: the name of the struct this function will be applied to
+//  unique: an arbitrary string to differentiate functions that should be applied
+//    to different structs that have the same name (e.g. two implementations of
+//    the "Employee" type.
+//  f: a function that gets called for all structs with this name. The function
+//    should verify the type of the struct and, if appropriate, return a non-empty
+//    string to be appended as the comment
+func RegisterHRSCommenter(typename, unique string, f HRSCommentFunc) {
```

Comment:
Just pass in a pointer (or if you follow my other comment) interface reference, and use that for identity, rather than `unique`.